enhancements/workload-partitioning/wide-availability-workload-partitioning.md

```diff
@@ -0,0 +1,731 @@
+	CPUPartitioningAllNodes CPUPartitioningMode = "AllNodes"
+)
+```
+
+### Machine Config Operator
+
+Once we have a global flag, and a way to set it at install time, we'll need to
+apply the needed configurations at install time during bootstrap. We will add
+this ability to MCO to generate and maintain the needed configurations before
+`kubelet` and the `api-server` stands up.
+
+We will add to the `kubelet` controller the ability to watch the
+`Infrastructure` resource and if CPU Partitioning is set to `AllNodes` we will
+generate the bootstrap and the controller will maintain the MCs from that point
+on. Things to note, this feature is explicitly designed to not be turned off, as
+such once set we will not remove the MCs.
+
+We will need to support upgrades for Single Node since this feature already
+exists for them but this implementation differs slightly. To avoid needless
+restarts, we will not alter the current configuration if we detect that we are
+in a single node topology and the nodes are already configured for CPU Partitioning.
+
+### Admission Controller
+
+We want to remove the checks in the admission controller which specifically
+verifies that partitioning is only applied to single node topology
+configuration. The design and configuration for any pod modification will remain
+the same, we simply will allow you to apply partitioning on non single node
+topologies.
+
+We will use the global identifier to correctly modify the pod spec with the
+`requests.cpu` for the new `requests[management.workload.openshift.io/cores]`
+that are used by the workload partitioning feature.
+
+However, for Single-Node we will continue to check the conventional way to be
+able to support the upgrade flow from 4.11 -> 4.12. After 4.12 release that
+logic should no longer be needed and will be removed.
+
+### Node Admission Plugin
+
+We want to ensure that nodes that are not setup for CPU Partitioning do not get
+added to a cluster that is designated for CPU Partitioning. We will create an
+admission plugin for Nodes that will validate if a cluster is setup for CPU
+Partitioning then the node MUST contain a Capacity resource for
+`workload.openshift.io/cores`. All node creation requests from Kubelet currently add
+that information on boot up when registering with the API Server, we will
+leverage that to ensure a CPU Partitioned cluster only contains nodes created
+for CPU Partitioning.
+
+We will also keep in mind upgrades from Single Node clusters which already
+contain this feature. During initial upgrade, Single Node clusters will not
+contain the CPUPartitioningMode, for this reason we will fall back to checking
+with the old logic for Single Node to insure we do not cause issues when
+upgrading. This check for Single Node should be something that happens on initial
+upgrades, as NTO will update the `Infra.Status.CPUPartitioningMode` to the correct
+value after initial boot.
```

Comment:
That shouldn't be the case because the fall back for the admission webhooks will still be present to check the nodes (the current way). The thinking behind this being only on initial start up is to avoid requests going off since the customers currently using this feature are very sensitive to increases in CPU usage. Since it's a one time check during this upgrade, it seemed appropriate.